data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/leader/LeaderScheduler.java

```diff
@@ -0,0 +1,172 @@
+package org.opensearch.dataprepper.plugins.mongo.leader;
+
+import org.opensearch.dataprepper.model.source.coordinator.enhanced.EnhancedSourceCoordinator;
+import org.opensearch.dataprepper.model.source.coordinator.enhanced.EnhancedSourcePartition;
+import org.opensearch.dataprepper.plugins.mongo.coordination.partition.ExportPartition;
+import org.opensearch.dataprepper.plugins.mongo.coordination.partition.GlobalState;
+import org.opensearch.dataprepper.plugins.mongo.coordination.partition.LeaderPartition;
+import org.opensearch.dataprepper.plugins.mongo.coordination.partition.StreamPartition;
+import org.opensearch.dataprepper.plugins.mongo.coordination.state.ExportProgressState;
+import org.opensearch.dataprepper.plugins.mongo.coordination.state.LeaderProgressState;
+import org.opensearch.dataprepper.plugins.mongo.coordination.state.StreamProgressState;
+import org.opensearch.dataprepper.plugins.mongo.configuration.CollectionConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+
+public class LeaderScheduler implements Runnable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LeaderScheduler.class);
+
+    /**
+     * Default duration to extend the timeout of lease
+     */
+    private static final int DEFAULT_EXTEND_LEASE_MINUTES = 3;
+
+    /**
+     * Default interval to run lease check and shard discovery
+     */
+    private static final Duration DEFAULT_LEASE_INTERVAL = Duration.ofMinutes(1);
+
+    private final List<CollectionConfig> collectionConfigs;
+
+    private final EnhancedSourceCoordinator coordinator;
+
+    private final Duration leaseInterval;
+
+    private LeaderPartition leaderPartition;
+
+    public LeaderScheduler(EnhancedSourceCoordinator coordinator, List<CollectionConfig> collectionConfigs) {
+        this(coordinator, collectionConfigs, DEFAULT_LEASE_INTERVAL);
+    }
+
+    LeaderScheduler(EnhancedSourceCoordinator coordinator,
+                    List<CollectionConfig> collectionConfigs,
+                    Duration leaseInterval) {
+        this.collectionConfigs = collectionConfigs;
+        this.coordinator = coordinator;
+        this.leaseInterval = leaseInterval;
+    }
+
+    @Override
+    public void run() {
+        LOG.info("Starting Leader Scheduler for initialization and stream discovery");
+
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                // Try to acquire the lease if not owned.
+                if (leaderPartition == null) {
+                    final Optional<EnhancedSourcePartition> sourcePartition = coordinator.acquireAvailablePartition(LeaderPartition.PARTITION_TYPE);
```

Comment:
Should we put a wait here between attempts to try to acquire the leader partition if it isn't acquired? Seems like an unnecessary amount of calls to source coordination if we call with no wait